File: driver/languages/typescript/parts/data_structures.ts
import { Scanner } from "./scanner";

export class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number = 0, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }
}

export class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val: number = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

export function buildList(size: number, sc: Scanner): ListNode | null {
    if (size === 0) return null;
    const head = new ListNode(sc.nextInt());
    let curr = head;
    for (let i = 1; i < size; i++) {
        curr.next = new ListNode(sc.nextInt());
        curr = curr.next;
    }
    return head;
}

export function buildTree(size: number, sc: Scanner): TreeNode | null {
    if (size === 0) return null;
    const vals: (number | null)[] = [];
    for(let i=0; i<size; i++) {
        const s = sc.nextString();
        vals.push(s === "null" ? null : parseInt(s!));
    }
    
    const root = new TreeNode(vals[0]!);
    const queue: TreeNode[] = [root];
    let i = 1;
    while (queue.length > 0 && i < size) {
        const node = queue.shift()!;
        if (vals[i] !== null) {
            node.left = new TreeNode(vals[i]!);
            queue.push(node.left);
        }
        i++;
        if (i < size && vals[i] !== null) {
            node.right = new TreeNode(vals[i]!);
            queue.push(node.right);
        }
        i++;
    }
    return root;
}

File: driver/languages/go/parts/data_structures.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func buildListNode(vals []int) *ListNode {
	var head *ListNode
	for i := len(vals) - 1; i >= 0; i-- {
		node := &ListNode{Val: vals[i], Next: head}
		head = node
	}
	return head
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTreeNode(vals []*int) *TreeNode {
	if len(vals) == 0 || vals[0] == nil {
		return nil
	}

	root := &TreeNode{Val: *vals[0]}
	queue := []*TreeNode{root}
	i := 1

	for len(queue) > 0 && i < len(vals) {
		curr := queue[0]
		queue = queue[1:]

		if i < len(vals) {
			if vals[i] != nil {
				curr.Left = &TreeNode{Val: *vals[i]}
				queue = append(queue, curr.Left)
			}
			i++
		}

		if i < len(vals) {
			if vals[i] != nil {
				curr.Right = &TreeNode{Val: *vals[i]}
				queue = append(queue, curr.Right)
			}
			i++
		}
	}

	return root
}
